lib/class-wp-block.php

```diff
@@ -0,0 +1,234 @@
+			$this->attributes = $block['attrs'];
+		}
+
+		if ( ! is_null( $this->block_type ) ) {
+			$this->attributes = $this->block_type->prepare_attributes_for_render( $this->attributes );
+		}
+
+		$this->available_context = $context;
+
+		if ( ! empty( $this->block_type->context ) ) {
+			foreach ( $this->block_type->context as $context_name ) {
+				if ( array_key_exists( $context_name, $this->available_context ) ) {
+					$this->context[ $context_name ] = $this->available_context[ $context_name ];
+				}
+			}
+		}
+
+		if ( ! empty( $block['innerBlocks'] ) ) {
+			$child_context = $this->available_context;
+
+			/* phpcs:disable WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase */
+			if ( ! empty( $this->block_type->providesContext ) ) {
+				foreach ( $this->block_type->providesContext as $context_name => $attribute_name ) {
+					if ( isset( $this->attributes[ $attribute_name ] ) ) {
+						$child_context[ $context_name ] = $this->attributes[ $attribute_name ];
+					}
+				}
+			}
+			/* phpcs:enable */
+
+			$this->inner_blocks = array_map(
+				function( $inner_block ) use ( $child_context, $registry ) {
+					return new WP_Block( $inner_block, $child_context, $registry );
+				},
+				$block['innerBlocks']
+			);
+		}
+
+		if ( ! empty( $block['innerHTML'] ) ) {
+			$this->inner_html = $block['innerHTML'];
+		}
+
+		if ( ! empty( $block['innerContent'] ) ) {
+			$this->inner_content = $block['innerContent'];
+		}
+	}
+
+	/**
+	 * Generates the render output for the block.
+	 *
+	 * @return string Rendered block output.
+	 */
+	public function render() {
+		global $post;
+
+		$is_dynamic    = $this->name && null !== $this->block_type && $this->block_type->is_dynamic();
+		$block_content = '';
+
+		$index = 0;
+		foreach ( $this->inner_content as $chunk ) {
+			$block_content .= is_string( $chunk ) ?
+				$chunk :
+				$this->inner_blocks[ $index++ ]->render();
+		}
+
+		if ( $is_dynamic ) {
+			$global_post   = $post;
+			$block_content = (string) call_user_func( $this->block_type->render_callback, $this, $block_content );
+			$post          = $global_post;
+		}
+
+		return $block_content;
+	}
+
+	/**
+	 * Returns true if an attribute exists by the specified attribute name, or
+	 * false otherwise.
+	 *
+	 * @link https://www.php.net/manual/en/arrayaccess.offsetexists.php
+	 *
+	 * @param string $attribute_name Name of attribute to check.
+	 *
+	 * @return bool Whether attribute exists.
+	 */
+	public function offsetExists( $attribute_name ) {
+		return isset( $this->attributes[ $attribute_name ] );
+	}
+
+	/**
+	 * Returns the value by the specified attribute name.
+	 *
+	 * @link https://www.php.net/manual/en/arrayaccess.offsetget.php
+	 *
+	 * @param string $attribute_name Name of attribute value to retrieve.
+	 *
+	 * @return mixed|null Attribute value if exists, or null.
+	 */
+	public function offsetGet( $attribute_name ) {
+		return isset( $this->attributes[ $attribute_name ] ) ?
```

Comment:
Accessing an undefined index in an array usually throws an `Undefined index` error in PHP, so I'm wondering if the implementation should do the something similar rather than guarding access with `isset`.

Would be good to see some unit tests for this ArrayAccess implementation.

Reply:
> Accessing an undefined index in an array usually throws an `Undefined index` error in PHP, so I'm wondering if the implementation should do the something similar rather than guarding access with `isset`.

Solid observation. For context, I adapted this more-or-less directly from the PHP documentation example:

https://www.php.net/manual/en/class.arrayaccess.php

But I do feel this should behave more or less like it would in the array form, especially considering our own backward-compatibility with providing this as an equivalent to `$block_attributes` array. In which case, I guess the remedy would be to remove the guard? And allow those warnings to happen as they would with any other improper array access?

> Would be good to see some unit tests for this ArrayAccess implementation.

Yep 👍 There's some small coverage in `test_passes_attributes_to_render_callback`, though only for "happy path" attribute access. An exhaustive set of tests covering the full set of these methods would be a good idea.

Reply:
> Accessing an undefined index in an array usually throws an `Undefined index` error in PHP, so I'm wondering if the implementation should do the something similar rather than guarding access with `isset`.
> 
> Would be good to see some unit tests for this ArrayAccess implementation.

Updated (including tests) in f8beb9e.